library/std/src/sys/hermit/condvar.rs

```diff
@@ -1,80 +1,108 @@
 use crate::ffi::c_void;
+use crate::mem::MaybeUninit;
 use crate::ptr;
-use crate::sync::atomic::{AtomicUsize, Ordering::SeqCst};
+use crate::sync::atomic::{
+    AtomicPtr, AtomicUsize,
+    Ordering::{Acquire, Release, SeqCst},
+};
 use crate::sys::hermit::abi;
 use crate::sys::locks::Mutex;
-use crate::sys_common::lazy_box::{LazyBox, LazyInit};
 use crate::time::Duration;
 
 // The implementation is inspired by Andrew D. Birrell's paper
 // "Implementing Condition Variables with Semaphores"
 
 pub struct Condvar {
     counter: AtomicUsize,
-    sem1: *const c_void,
-    sem2: *const c_void,
+    sem1: AtomicPtr<c_void>,
+    sem2: AtomicPtr<c_void>,
 }
 
-pub(crate) type MovableCondvar = LazyBox<Condvar>;
+pub(crate) type MovableCondvar = Condvar;
 
-impl LazyInit for Condvar {
-    fn init() -> Box<Self> {
-        Box::new(Self::new())
+#[cold]
+fn init_semaphore(sem: &AtomicPtr<c_void>) -> *mut c_void {
+    let new = unsafe {
+        let mut new = MaybeUninit::uninit();
+        let _ = abi::sem_init(new.as_mut_ptr(), 0);
+        new.assume_init() as *mut c_void
+    };
+
+    match sem.compare_exchange(ptr::null_mut(), new, Release, Acquire) {
+        Ok(_) => new,
+        Err(sem) => unsafe {
+            let _ = abi::sem_destroy(new);
+            sem
+        },
     }
 }
 
-unsafe impl Send for Condvar {}
-unsafe impl Sync for Condvar {}
-
 impl Condvar {
-    pub fn new() -> Self {
-        let mut condvar =
-            Self { counter: AtomicUsize::new(0), sem1: ptr::null(), sem2: ptr::null() };
-        unsafe {
-            let _ = abi::sem_init(&mut condvar.sem1, 0);
-            let _ = abi::sem_init(&mut condvar.sem2, 0);
+    #[inline]
+    pub const fn new() -> Self {
+        Self {
+            counter: AtomicUsize::new(0),
+            sem1: AtomicPtr::new(ptr::null_mut()),
+            sem2: AtomicPtr::new(ptr::null_mut()),
         }
-        condvar
+    }
+
+    #[inline]
+    fn semaphores(&self) -> (*const c_void, *const c_void) {
+        let mut sem1 = self.sem1.load(Acquire);
+        if sem1.is_null() {
+            sem1 = init_semaphore(&self.sem1);
+        }
+
+        let mut sem2 = self.sem2.load(Acquire);
+        if sem2.is_null() {
+            sem2 = init_semaphore(&self.sem2);
+        }
+
+        (sem1, sem2)
     }
 
     pub unsafe fn notify_one(&self) {
         if self.counter.load(SeqCst) > 0 {
             self.counter.fetch_sub(1, SeqCst);
```

Comment:
This isn't part of the code you've changed, but isn't this a race condition?

Reply:
Yes it is. There are other issues: if `wait_timeout` times out, `notify_one` does not wake up any thread.